src/main/java/in/projecteka/consentmanager/user/OtpAttemptService.java

```diff
@@ -0,0 +1,95 @@
+package in.projecteka.consentmanager.user;
+
+import in.projecteka.consentmanager.clients.ClientError;
+import in.projecteka.consentmanager.user.model.OtpAttempt;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+import reactor.core.publisher.Mono;
+
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.List;
+
+@Service
+@AllArgsConstructor
+public class OtpAttemptService {
+
+    private final OtpAttemptRepository otpAttemptRepository;
+    private final UserServiceProperties userServiceProperties;
+
+    public Mono<Void> validateOTPRequest(String identifierType, String identifierValue, OtpAttempt.Action action, String cmId) {
+        OtpAttempt.OtpAttemptBuilder builder = OtpAttempt.builder()
+                .identifierType(identifierType)
+                .identifierValue(identifierValue)
+                .action(action)
+                .cmId(cmId);
+        return otpAttemptRepository.getOtpAttempts(builder.build(), userServiceProperties.getMaxOtpAttempts())
+                .filter(otpAttempts -> otpAttempts.size() == userServiceProperties.getMaxOtpAttempts())
+                .filter(this::isNoneBlockedExceptLatest)
+                .flatMap(this::validateLatestAttempt)
+                .flatMap(this::validateAttemptsLimit)
+                .switchIfEmpty(Mono.defer(() -> saveOTPAttempt(builder.attemptStatus(OtpAttempt.AttemptStatus.SUCCESS).build())));
+    }
+
+    public Mono<Void> validateOTPRequest(String identifierType, String identifierValue, OtpAttempt.Action action) {
+        return validateOTPRequest(identifierType, identifierValue, action, "");
+    }
+
+    public Mono<Void> validateOTPSubmission(OtpAttempt otpAttempt) {
+        return otpAttemptRepository.getOtpAttempts(otpAttempt, userServiceProperties.getOtpMaxInvalidAttempts())
+                .filter(otpRequestAttempts -> otpRequestAttempts.size() == userServiceProperties.getOtpMaxInvalidAttempts())
+                .flatMap(otpRequestAttempts -> {
+                    OtpAttempt latestAttempt = otpRequestAttempts.stream().findFirst().get();
+                    boolean isBlocked = isWithinTimeLimit(latestAttempt, userServiceProperties.getOtpInvalidAttemptsBlockPeriodInMin());
+                    return isBlocked ? Mono.error(ClientError.tooManyInvalidOtpAttempts()) : removeMatchingAttempts(otpAttempt).then(Mono.empty());
+                });
+    }
+
+    public Mono<Void> removeMatchingAttempts(OtpAttempt otpAttempt) {
```

Comment:
do we really need this methods..can't we directly call the method of repository?

Reply:
I think it's better not to call the repository methods directly. Later on if we want to change the logic in `removeMatchingAttempts` then we will have to change only in service instead of changing it everywhere.

Reply:
ok